### PR TITLE
SIMD: Add sum intrinsics for float/double.

### PR DESCRIPTION
**Pull request by Qiyu8**
The [origin PR](https://github.com/numpy/numpy/pull/17049) is too large to review, So I will split to several small PRs, Here is the sum intrinsics that was about to used in `einsum`, The intrinsics has been fully discussed and tested.

### CODE REVIEW

**Comment by mattip** [comment]
@seiko2plus looks ok?

File: numpy/core/src/common/simd/neon/arithmetic.h [hunk 1]
```diff
@@ -119,3 +119,16 @@
...
+{
+    float32x2_t r = vadd_f32(vget_high_f32(a), vget_low_f32(a));
+    return vget_lane_f32(vpadd_f32(r, r), 0);
+}
+#ifdef __aarch64__
+    NPY_FINLINE double npyv_sum_f64(float64x2_t a)
+    {
+        return vget_lane_f64(vget_low_f64(a) + vget_high_f64(a), 0);
+    }
+#endif
```
Comment (seiko2plus) [inline numpy/core/src/common/simd/neon/arithmetic.h:134]:
```suggestion
// Horizontal add: Calculates the sum of all vector elements.
#if NPY_SIMD_F64
    #define npyv_sum_f32 vaddvq_f32
    #define npyv_sum_f64 vaddvq_f64
#else
    NPY_FINLINE float npyv_sum_f32(npyv_f32 a)
    {
        npyv_f32 r = vadd_f32(vget_high_f32(a), vget_low_f32(a));
        return vget_lane_f32(vpadd_f32(r, r), 0);
    }
#endif
```

File: numpy/core/src/common/simd/vsx/arithmetic.h [hunk 2]
```diff
@@ -117,3 +117,15 @@
 #define npyv_nmulsub_f64 vec_nmadd
 
 #endif // _NPY_SIMD_VSX_ARITHMETIC_H
+
+// Horizontal add: Calculates the sum of all vector elements.
+NPY_FINLINE float npyv_sum_f32(npyv_f32 a)
+{
+    return vec_extract(a, 0) + vec_extract(a, 1) +
+    vec_extract(a, 2) + vec_extract(a, 3);
+}
```
Comment (seiko2plus) [inline numpy/core/src/common/simd/vsx/arithmetic.h:126]:
```suggestion
NPY_FINLINE float npyv_sum_f32(npyv_f32 a)
{
    npyv_f32 sum = vec_add(a, vec_mergeh(a, a));
    return vec_extract(sum, 0) + vec_extract(sum, 1);
}
```

File: numpy/core/src/common/simd/avx2/arithmetic.h [hunk 3]
```diff
@@ -117,3 +117,23 @@
     }
 #endif // !NPY_HAVE_FMA3
 #endif // _NPY_SIMD_AVX2_ARITHMETIC_H
+
+// Horizontal add: Calculates the sum of all vector elements.
```
Comment (seiko2plus) [inline numpy/core/src/common/simd/avx2/arithmetic.h:121]: please, move the intrinsics inside the header guard `_NPY_SIMD_AVX2_ARITHMETIC_H`,
same thing for other SIMD extensions.

Comment (seiko2plus) [inline numpy/core/src/common/simd/avx2/arithmetic.h:121]: please, move the intrinsics inside the header guard `_NPY_SIMD_AVX2_ARITHMETIC_H`,
same thing for other SIMD extensions.

**Comment by seiko2plus** [comment]
I prefer adding a testing unit for any new intrinsics to keep things under control.
@Qiyu8 please could you add a testing case for them?
If yes, then you have to add new methods for the new intrinsics within [_simd.dispatch.c.src](https://github.com/numpy/numpy/blob/master/numpy/core/src/_simd/_simd.dispatch.c.src).  for example:
```C
// try to follow the current defentions in the way of sorting the source
// this how you should define the new python methods
SIMD_IMPL_INTRIN_1(sum_f32, f32, vf32)
SIMD_IMPL_INTRIN_1(sum_f64, f64, v64)
// and that how we attach them
SIMD_INTRIN_DEF(sum_f32)
SIMD_INTRIN_DEF(sum_f64)
```
Once you get done bringing the new methods to `_simd` module, go to [test_simd.py](https://github.com/numpy/numpy/blob/master/numpy/core/tests/test_simd.py) and add testing cases for them. for example:

```Python
def test_reduce(self):
    data = self._data()
    vdata = self.load(self._data())
    # reduce sum
    data_sum = sum(data)
    vsum = self.sum(vdata)
    assert vsum == data_sum
````

You can also have some fun and try to use `_simd` module as tool for designing and discovering, for example
```Python
# 1- bring the baseline via dict `targets` or attribute `baseline`
from numpy.core._simd import targets, baseline
npyv = targets["targets"] # or baseline
# you can also dump `targets()` to get the supported SIMD extensions
# by default build option `--test-simd` contains the most common SIMD extentions
if not npyv.simd: # equivalent to C def `NPY_SIMD`
    print((
        "How that possible? changed the default build settings?\n"
        "maybe you running it under armhf, then get targets['NEON']"
    ))
    return

a = npyv.load_f32(range(npyv.nlanes_f32))
print(npyv.sum_f32(a))
````

**Comment by Qiyu8** [comment]
@seiko2plus Thanks for your detailed recommendations, The `_simd.dispatch.c.src` provides perfect functional test interface for universal intrinsics, we should write a doc such as `how to write a simd test` because the additional complexity compared to regular apis.

**Review by seiko2plus — APPROVED** [review]
LGTM, Thank you.

File: numpy/core/tests/test_simd.py [hunk 4]
```diff
@@ -512,6 +512,16 @@ def test_arithmetic_div(self):
...
+    def test_arithmetic_reduce_sum(self):
+        if not self._is_fp():
+            return
+        # reduce sum
+        data = self._data()
+        vdata = self.load(data)
+
+        data_sum = sum(data)
+        vsum = self.sum(vdata)
+        assert vsum == data_sum
```
Comment (mattip) [inline numpy/core/tests/test_simd.py:524]: nice!

**Comment by mattip** [comment]
Thanks @Qiyu8

**Pull request merged**